### PR TITLE
[RFC] hooks: add support for `.test` files and add some initial tests

### PR DESCRIPTION
**Pull request by mvo5**
The current hooks do not have a good testing story. To fix this
this PR adds support for "hook/foo.test" files that will be run
at the end of the prime phase during the snap build.

It also adds two initial tests and converts the existing ABI test.
If this looks reasonable we can write tests for the remaining hooks
to ensure we do not regress again.

### CODE REVIEW

File: snapcraft.yaml [hunk 1]
```diff
@@ -26,3 +26,5 @@ parts:
       unset LD_LIBRARY_PATH;
       export PATH="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin";
       snapcraftctl prime
+      # ensure buildin tests are run
```
Comment (zyga) [inline snapcraft.yaml]:
```suggestion
      # ensure built-in tests are run
```

File: hooks/003-remove-dbus-machine-id.test [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    exit 1
+fi
+
+if [ ! -e /etc/machine-id ]; then
+    echo "/etc/machine-id must exit"
+    exit 1
+fi
+
+if [ -s /etc/machine-id ]; then
+    echo "/etc/machine-id must exit but must be empty"
```
Comment (zyga) [inline hooks/003-remove-dbus-machine-id.test]:
```suggestion
    echo "/etc/machine-id exists but must be empty"
```

File: hooks/003-remove-dbus-machine-id.test [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+set -e
+
+if [ -f /var/lib/dbus/machine-id ]; then
+    echo "cannot have /var/lib/dbus/machine-id in the image"
+    exit 1
+fi
+
+if [ ! -e /etc/machine-id ]; then
+    echo "/etc/machine-id must exit"
```
Comment (zyga) [inline hooks/003-remove-dbus-machine-id.test]:
```suggestion
    echo "/etc/machine-id must exist"
```

File: README.md [hunk 4]
```diff
@@ -10,6 +10,17 @@ To build this snap locally you need snapcraft. The project must be built as real
...
 ```
 
+# Writing code
+
+The usual way to add functionality is to write a shell script hook
+with the `.chroot` extenstion under the `hooks/` directory. These hooks
+are run inside the base image filesystem.
+
+Each hook should have a matching `.test` file that validates that the
+`.chroot` file works as expected. Those `.test` scripts will be run
```
Comment (pedronis) [inline README.md]: it is not very clear from this description that these tests will be run after building with snapcraft,  if I understand the travis.yml.

Also I wonder whether "sanity" or "check" would be a better name for them and this process.

**Review by sil2100 — COMMENTED** [review]
Ok, this looks fine (besides those small typos as highlighted by zyga).

Personally (that's just a preference though) I don't like the idea of having both the hooks and the tests in the same directory. It might get a bit confusing at first glance, since looking at the hooks/ directory you'll see files that are not actual hooks. What I would prefer personally is moving those tests to a separate hook-tests/ or similar directory, still following the same layout for running them as here.

But that's just my 5 cents.

File: README.md (hunk 4 above)
Comment (mvo5) [reply README.md]: Thanks! I updated the documentation now. As for `test` vs `check` vs `sanity` I have no super strong opinion. Those are not tests in the traditional sense so a rename does make sense. I feel the upside of using `tests` is that people who write code may look for "tests" when wanting to write tests more readily than looking for `check` or `sanity`.

**Comment by mvo5** [comment]
@sil2100 Thank you! I moved the test files out now, the original "design" was inspired by go (which mixes source and test in a very similar way) but as foundations own this I'm fine accommodating your preferences :)

**Review by sil2100 — APPROVED** [review]

**Pull request merged**